### PR TITLE
Ticket sellers in Kraków

### PR DESCRIPTION
**Pull request by matkoniecz**
I have some pending TODOs that I will look into but maybe these are obviously fixable so if anyone knows how to solve them...

How can I stop NSI matcher from running on some cases? 

Not everything is vending machine (or should I report to NSI that they should have also offices listed?)

Maybe I should skip offices and process them in a separate spider? But it will pointlessly query twice.

### CODE REVIEW

**Comment by matkoniecz** [comment]
"In the cases where it fails for some reason then the automation can be disabled for the spider" in https://github.com/alltheplaces/alltheplaces/blob/master/docs/WIKIDATA.mdoffers tantalizing view that it can be done.

Still no idea how to do this.

**Comment by CloCkWeRX** [comment]
Which bit is failing/you want to ignore the automation on?

```
class HipermaxiBOSpider(AgileStoreLocatorSpider):
    name = "hipermaxi_bo"
    item_attributes = {
        "brand_wikidata": "Q81968262", # Override wikidata match spider wide
        "brand": "Hipermaxi", 
    }
```

or
```
def parse_item(self, item, location):
```
Is how you can generally do the specific decoration of an item from whatever you are parsing, then "yield" which is _kind of_ like a 

```return variable_name```

but it's actually kind of like

```return super(variable_name)```

(Apologies if you already know that bit, just assuming you are more used to java influenced languages... though looks like kotlin has the concept in an iteration context)

File: locations/spiders/kkm_vending_krk_pl.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+            TICKET_OFFICE_VALUES = ["2"]
+            TICKET_MACHINE_WITH_INFO_KIOSK_VALUES = ["13"]
+            TICKET_MACHINE_VALUES = ["1", "9"]
+            TICKET_MACHINE_VALUES_WITH_NO_CACH_SUPPORT = ["41"]
+            if location["TypeId"] in TICKET_OFFICE_VALUES:
+                apply_category(Categories.SHOP_TICKET, item)
+                item["extras"]["ticket"] = "public_transport"
+                item["name"] = "Punkt Obsługi Pasażerów KMK"
+                # TODO: stop matching on NSI here - these are offices
```
Comment (CloCkWeRX) [inline locations/spiders/kkm_vending_krk_pl.py:50]: https://github.com/alltheplaces/alltheplaces/blob/master/locations/pipelines/apply_nsi_categories.py#L10

**Comment by matkoniecz** [comment]
https://github.com/alltheplaces/alltheplaces/pull/7610/files#diff-ebabc5f645c0760030ed6540d6f5f71b7ef7cc4d527da503cd10fc1593d76286R46 branch also gets tags derived from https://github.com/alltheplaces/alltheplaces/pull/7610/files#diff-ebabc5f645c0760030ed6540d6f5f71b7ef7cc4d527da503cd10fc1593d76286R15

I would want to disable NSI pipeline within `if location["TypeId"] in TICKET_OFFICE_VALUES:` if

File: locations/spiders/kkm_vending_krk_pl.py (hunk 1 above)
Comment (matkoniecz) [reply locations/spiders/kkm_vending_krk_pl.py:50]: If I look at it right then presence of `nsi_id` with any value will terminate matching. I will try that.

Comment (CloCkWeRX) [reply locations/spiders/kkm_vending_krk_pl.py:50]: Option 1:
https://github.com/alltheplaces/alltheplaces/blob/4d5453f71f9d57db4969b9c77998ce054470878c/locations/spiders/argos.py#L18

post process_item, then undo whatever it did when you want to.

**Comment by Cj-Malone** [comment]
> or should I report to NSI that they should have also offices listed

If there is a decent amount, yes it should be in NSI. The "hard" min is 10 POIs for NSI, however my personal min is a bit higher.

File: locations/spiders/kkm_vending_krk_pl.py (hunk 1 above)
Comment (matkoniecz) [reply locations/spiders/kkm_vending_krk_pl.py]: https://github.com/alltheplaces/alltheplaces/issues/6907#issuecomment-1966174916 solves this as far as I see

thanks for all help!

**Comment by matkoniecz** [comment]
> If there is a decent amount, yes it should be in NSI. The "hard" min is 10 POIs for NSI, however my personal min is a bit higher.

There is less than then that - three locations, and it is city specific so number unlikely to grow much

**Comment by matkoniecz** [comment]
So what remains now is either implementing opening hours parsing for their syntax or giving up on opening hours.

(and then looking at data again)

File: locations/spiders/kkm_vending_krk_pl.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            item["ref"] = location["Id"]
+            item["lat"] = location["Latitude"]
+            item["lon"] = location["Longitude"]
+            item["extras"]["location_description"] = location["Location"]
+            if location["OpeningHours"] == "Całą dobę":
+                item["extras"]["opening_hours"] = "24/7"
+            else:
+                entry = OpeningHours()
+                entry.add_ranges_from_string(location["OpeningHours"], DAYS_PL)
+                item["extras"]["opening_hours"] = entry.as_opening_hours()
```
Comment (Cj-Malone) [inline locations/spiders/kkm_vending_krk_pl.py:31]:
```suggestion
                item["opening_hours"] = "24/7"
            else:
                item["opening_hours"] = OpeningHours()
                item["opening_hours"].add_ranges_from_string(location["OpeningHours"], DAYS_PL)
```

`opening_hours` isn't an extra, it's a direct property. Also if we output the `OpeningHours` object, we skip validation if it's the string (`as_opening_hours`) we do some regex on it.

File: locations/spiders/kkm_vending_krk_pl.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                item["nsi_id"] = "N/A"
+                yield item
+            if location["TypeId"] in TICKET_MACHINE_WITH_INFO_KIOSK_VALUES + TICKET_MACHINE_VALUES:
+                item["extras"]["payment:cash"] = "yes"
+                item["extras"]["payment:cards"] = "yes"
+                yield item
+            if location["TypeId"] in TICKET_MACHINE_VALUES_WITH_NO_CACH_SUPPORT:
+                item["extras"]["payment:cash"] = "no"
+                item["extras"]["payment:cards"] = "yes"
+                yield item
```
[the remark this answers is not in the document]
Comment (Cj-Malone) [reply locations/spiders/kkm_vending_krk_pl.py:49]: (Un tested)

[the remark this answers is not in the document]
Comment (matkoniecz) [reply locations/spiders/kkm_vending_krk_pl.py:49]: thanks, I wanted to log missing types and forgot to add this!

**Review by Cj-Malone — APPROVED** [review]
LGTM, thanks.
https://alltheplaces-data.openaddresses.io/map.html?show=https://alltheplaces-data.openaddresses.io/ci/8063707830/kkm_vending_krk_pl/output.geojson

**Pull request merged**